go-controller/pkg/clustermanager/networkconnect/hybrid_connect_subnet_allocator.go

```diff
@@ -0,0 +1,351 @@
+	hca.Lock()
+	defer hca.Unlock()
+
+	if hca.layer3Allocator == nil {
+		return fmt.Errorf("allocator not initialized")
+	}
+
+	// Try efficient batch marking in Layer3 allocator first
+	if err := hca.layer3Allocator.MarkAllocatedNetworks(owner, subnets...); err == nil {
+		// All subnets were successfully marked in Layer3 - store all subnets
+		if len(subnets) > 0 {
+			hca.allAllocations[owner] = append(hca.allAllocations[owner], subnets...)
+		}
+		return nil
+	}
+
+	// If batch marking failed, try marking each subnet individually
+	// This handles cases where some subnets are Layer3 and others are Layer2
+	for _, subnet := range subnets {
+		if err := hca.markSingleAllocatedSubnet(owner, subnet); err != nil {
+			return fmt.Errorf("failed to mark subnet %s: %v", subnet.String(), err)
+		}
+	}
+
+	return nil
+}
+
+// addAllocation adds a subnet to an owner's allocation list
+func (hca *hybridConnectSubnetAllocator) addAllocation(owner string, subnet *net.IPNet) {
+	hca.allAllocations[owner] = append(hca.allAllocations[owner], subnet)
+}
+
+// markSingleAllocatedSubnet is the implementation for marking a single subnet
+// This is used by both MarkAllocatedSubnet and MarkAllocatedSubnets
+func (hca *hybridConnectSubnetAllocator) markSingleAllocatedSubnet(owner string, subnet *net.IPNet) error {
+	// Try to mark in Layer3 allocator first
+	if err := hca.layer3Allocator.MarkAllocatedNetworks(owner, subnet); err == nil {
+		hca.addAllocation(owner, subnet)
+		return nil
+	}
+
+	// If that fails, it might be a Layer2 subnet in one of our pools
+	// Determine if this is IPv4 or IPv6
+	isIPv6 := subnet.IP.To4() == nil
+
+	// Check all pools for this subnet
+	for _, pool := range hca.layer2Pools {
+		var poolCIDR *net.IPNet
+
+		// Select the appropriate pool CIDR based on address family
+		if isIPv6 {
+			poolCIDR = pool.v6PoolCIDR
+		} else {
+			poolCIDR = pool.v4PoolCIDR
+		}
+
+		// Check if this pool contains this subnet
+		if poolCIDR != nil && poolCIDR.Contains(subnet.IP) {
+			// This pool should contain this subnet - use the dual-stack allocator
+			if err := pool.allocator.MarkAllocatedNetworks(owner, subnet); err == nil {
+				pool.allocated[owner] = append(pool.allocated[owner], subnet)
+				hca.addAllocation(owner, subnet)
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("subnet %s does not belong to any known range", subnet.String())
+}
+
+// GetAllOwners returns the list of all owners that have allocations
+func (hca *hybridConnectSubnetAllocator) GetAllOwners() []string {
+	hca.RLock()
+	defer hca.RUnlock()
+
+	owners := make([]string, 0, len(hca.allAllocations))
+	for owner := range hca.allAllocations {
+		owners = append(owners, owner)
+	}
+	return owners
+}
+
+// ReleaseAllAllocations releases all allocations in this allocator
+// This should be called when deleting a CNC
+func (hca *hybridConnectSubnetAllocator) ReleaseAllAllocations() {
```

Comment:
is this actually needed? if the CNC is being deleted, then I guess the whole `hybridConnectSubnetAllocator` will be deleted anyway?